apps/api_web/lib/api_web.ex

```diff
@@ -19,8 +19,8 @@ defmodule ApiWeb do
       # Start the endpoint when the application starts
       worker(ApiWeb.RateLimiter, []),
       worker(RequestTrack, [[name: ApiWeb.RequestTrack]]),
-      supervisor(ApiWeb.EventStream.Supervisor, []),
-      supervisor(ApiWeb.Endpoint, [])
+      supervisor(ApiWeb.Endpoint, []),
+      supervisor(ApiWeb.EventStream.Supervisor, [])
```

Comment:
In my testing, doing it this way seemed necessary to get the connections closed cleanly — if the endpoint shuts down first, the event-stream processes are terminated without knowing they're supposed to finish up, since the DiffServer is what tells them that, but _it_ doesn't know it's getting shut down until after the Endpoint is shut down and it's too late. But I hadn't thought about the dependency when things are starting up, and you're definitely right.

Possibly the solution involves what you mentioned above:

> Did you look at using the `drainer` configuration from the Cowboy2Adapter?

I hadn't seen this before now, and it seems like it might help but is also very under-documented... I can't find any guidance on how to use it or what it can do (a not-yet-published [version of the doc](https://github.com/phoenixframework/phoenix/blob/c8883af5582a38496e4b7e45e05d3a4d759a6caa/lib/phoenix/endpoint/cowboy2_adapter.ex#L17-L21) at least clarifies what the _default_ value does, but to me this only raises more questions). Have you used it before?

Reply:
I haven't used it before. The code in https://github.com/elixir-plug/plug_cowboy/blob/v2.4.1/lib/plug/cowboy/drainer.ex  is interesting, though. It's a separate process, running after the Cowboy listeners, which waits for them to stop having any open connections.

For us, it sounds like what we want is for `DiffServer` to remain before `Endpoint`, and for a new server to be after `Endpoint` and be responsible for shutting down the long-running connections, similar to what `Plug.Cowboy.Drainer` is doing here.

Reply:
Yes, I'm not sure if the `drainer` config can help us here; it seems like the mentioned "drainer process" is [always `Plug.Cowboy.Drainer`](https://github.com/phoenixframework/phoenix/blob/c8883af5582a38496e4b7e45e05d3a4d759a6caa/lib/phoenix/endpoint/cowboy2_adapter.ex#L71), and the config might be better named `drainer_opts`.

> For us, it sounds like what we want is for `DiffServer` to remain before `Endpoint`, and for a new server to be after `Endpoint` and be responsible for shutting down the long-running connections

Hmmmmm... I'm trying to think of how we do this without races. So far I'm thinking something like this (calling the new process `Canary`):

```
ApiWeb.Supervisor => EventStream.Supervisor => ConnectionRegistry
                  => Endpoint                        |
                  => EventStream.Canary -------------/
```

New event-stream connections register themselves in a registry owned by the EventStream supervisor (since it has to be running before the Endpoint). When the canary is shut down, it sends all running event-stream connections a message using the registry. It then shuts down immediately, but because of the Cowboy drainer, the connections will have a few seconds to respond to the message and finish up before the Endpoint shuts down.

The race I can see here is that there's a tiny window _after_ the Canary has fetched the list of event-stream connections, but _before_ the Endpoint starts draining, where a new event-stream connection could be established. It wouldn't know about the shutdown, and so would still experience the problem we're trying to fix. Maybe this would be uncommon in practice, but it bugs me that it would be possible. (it also bugs me how convoluted this is, but maybe it's just a convoluted problem) What do you think?

Reply:
> The race I can see here is that there's a tiny window _after_ the Canary has fetched the list of event-stream connections, but _before_ the Endpoint starts draining, where a new event-stream connection could be opened.

Hasn't the load balancer already dropped us at this point? I don't think it's possible for there to be a new connection by the time we're disconnecting existing connections.

Reply:
Oh, you're right, it would have. So this race might exist in a non-AWS context, but fortunately we don't have to worry about that. 😅 If you think the approach is good otherwise, I will give it a shot.

Reply:
Yeah, I'd go for that rather than re-ordering the supervision tree.